pep_sphinx_extensions/pep_theme/static/sticky_banner.js

```diff
@@ -0,0 +1,28 @@
+"use strict";
+
+// Inject a style element into the document head that adds scroll-margin-top to
+// all elements with an id attribute. This is used to offset the scroll position
+// when clicking on a link to an element with an id attribute. The offset is
+// equal to the height of the sticky banner.
+document.addEventListener("DOMContentLoaded", () => {
+  const stickyBanners = document.getElementsByClassName("sticky-banner");
+  if (!stickyBanners.length) {
+    return;
+  }
+
+  const stickyBanner = stickyBanners[0];
+  const node = document.createElement("style");
+  node.id = "sticky-banner-style";
+  document.head.appendChild(node);
+
+  function adjustBannerMargin() {
+    const text = document.createTextNode(
+      ":target { scroll-margin-top: " + stickyBanner.offsetHeight + "px; }"
+    );
+    node.replaceChildren(text);
+  }
+
+  adjustBannerMargin();
+  document.addEventListener("resize", adjustBannerMargin);
+  document.addEventListener("load", adjustBannerMargin);
```

Comment:
I still occasionally see the previous issue, the banner covering the headings on uncached load, on both FF on Windows and Android (Pixel 3a), though less often then before. However, I'm seeing a new issue consistently on the latter—the scroll height is several lines off in the other direction on uncached load, presumably because either there is a resize on `load` after scrolling occurs, or scrolling just occurs too early for any of the events to fire. As mentioned above, the adapting the aforementioned snippet should properly fix both those issues (and any other similar ones) with no obvious downsides:

```suggestion
  function adjustBannerMarginAndScroll {
    adjustBannerMargin();
    if (location.hash.length > 1) {
      document.getElementById(location.hash.substring(1)).scrollIntoView();
    }
  }

  adjustBannerMarginAndScroll();
  document.addEventListener("load", adjustBannerMarginAndScroll);
  document.addEventListener("resize", adjustBannerMargin);
```

Reply:
> document.getElementById(location.hash.substring(1)).scrollIntoView();

I'm wary of adding this, since this is basically fiddling with default browser scroll behaviours. From my (not vast) experience in web development, one of the things I've picked up on as a general tip is "don't mess with scroll".

> uncached load

Do you think improving this is a blocking concern for this, or is this something we can file an issue for and improve in a follow up?

Reply:
> I'm wary of adding this, since this is basically fiddling with default browser scroll behaviours. From my (not vast) experience in web development, one of the things I've picked up on as a general tip is "don't mess with scroll".

I'd be more wary if we were trying to manually override the default fragment id to scroll to and/or manually dialing-in a scroll offset, both of which I did on spyder-ide/spyder-docs#207 (which took a bit of finagling, particularly since I couldn't necessarily rely on ``scroll-margin`` for the latter back then—its quite the bunch of hackery, though it seems to works fairly reliably in practice). However, in this case, all we're doing is (re-)issuing a basic scroll command to the same fragment id that the browser has or will have navigated to.

Furthermore, _**we already are**_ fiddling with the browser's scroll behavior by dynamically modifying the `scroll-margin` in JS during page rendering, which races the browser's own default scroll and causes the multiple real-world edge-case bugs reported here (whereas a basic scroll to the same existing target is much less fiddly). Given the solution here specifically fixes those outstanding defects, whereas it is unclear by what plausible theoretical mechanism it could introduce new hypothetical ones, I'm having trouble seeing how this is a net increase in risk.

> Do you think improving this is a blocking concern for this, or is this something we can file an issue for and improve in a follow up?

It's not a huge blocking concern, but as we've already put in the time to implement and test a working solution to multiple actual reported issues with this PR that can be applied right now, IMO it would make the most sense to just apply it here, merge and iterate from there if further changes are needed.

That being said, if there are in fact practical real-world problems with this solution such that it can't be applied here, realistically the cost-benefit of trying to work around that is likely such that it's not worth spending any of our additional time on a dedicated followup, so if that is in fact the case I would just drop it and move on to more important things :)

Reply:
I think it's worth spending the effort of getting this right _and_ that what we have at the moment is good-enough for a first cut -- to that end, I have filed #3065 for tracking the proper fix.

Reply:
In hindsight, it was indeed a mistake for me to waste everyone's time reporting and then persist with this issue when it really wasn't worth everyone's trouble, and for that I apologize.

As mentioned, I'm not sure what a "proper fix" beyond this would plausibly look like, and I'm not sure it's really worth any more of our valuable time going beyond "good enough" for what is already a seemingly rare and difficult to repro corner case, but maybe my JS-fu is just too limited to even envision it given your much greater experience in this area.

I regret that I seem to have encouraged you into spending it on one small issue in our bespoke one-off theme, but it would be wrong of me to tell you how to spend your free time, of course, and if that's really what you want to do, given it's my fault the least I can do is help test whatever you come up with.

Reply:
And I'm sorry I got a little frustrated; to be clear I'm frustrated at myself for wasting other people's time (most notably yours) after all the hard work you already put into this issue, and will definitely keep this in mind for the future.